### PR TITLE
Initial version of stack clash protection post

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
Still a WIP, needs a last step from @sylvestre

**Comment by lattner** [comment]
Looking great so far, thank you for working on this.  Please lemme know when you're content complete and I'll happily provide a pass of feedback

**Comment by serge-sans-paille** [comment]
@lattner ready for review! I did a simple markdown rendering, but didn't check Hugo integration, esp. the reference section.

**Comment by sylvestre** [comment]
Many thanks @serge-sans-paille ! :+1:

**Comment by serge-sans-paille** [comment]
@lattner gentle ping  ;-)

**Comment by serge-sans-paille** [comment]
@sqlbyme I assume you deleted that PR because of the `master` → `main` renaming (?) I'll update a new PR to `main` then.

**Comment by sqlbyme** [comment]
Hi,
My apologies, I did not actually delete your pull request, rather I did not
intend to.  I believe when I deleted the master branch, GitHub
automatically closed your PR.  Sorry for the confusion.

Thanks,
Mike

**Comment by lattner** [comment]
i'm sorry for the delay getting to this but I'm really happy to review this this week, could you create another PR?

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+has even been identified as a blocker by Fedora to move from gcc to clang as the
+compiler for some projects that already made the move upstream, leading to [extra
+maintenance][FedoraClangGCC] for packagers.
+
+Support for this flag [landed in Clang in 2020][CLANGStackClashProtection],
+only for X86, SystemZ and PowerPC. Its implementation is a result of a fruitful
+collaboration between LLVM, Firefox and Rust developpers.
+
+
+Rust already had a countermeasure implemnted in the form of a runtime call to
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:31]: typo implemnted

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+The probing strategy for stack allocation varies based on the size of the stack
+allocation. If it's smaller than `PAGE_SIZE`, thanks to (2) no probing is
+needing. If it's below a small multiple of ``PAGE_SIZE``, then the probing loop
+can be unrolled. Otherwise a probing loop alternates stack allocation of
+`PAGE_SIZE` bytes and probe, starting with the allocation thanks to (1).
+
+As side effect of (2) is that when performing a dynamic allocation, we need to
+probe *before* updating the stack, otherwise we got a hole in the protection.
+This probe cannot be done after the stack update, even with an offset, because
+of (3). Otherwise we end up with a bug as this one found in gcc [GCCProbeBug]_
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:63]: I'd recommend normalizing "gcc" to "GCC" unless referring to the command line tool.

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+## Performance Testing
+
+Over the years, Mozilla has developed a few tools to evaluate performance
+impact of changes, from micro-benchmark to page loads. These tools have been key
+to improve Firefox overall performances but also evaluate the impact of the [move
+to Clang][move-clang] on all platforms done a couple years ago.
+
+The usual procedure to evaluate performances improvements/regressions is to:
+
+1. Run two builds with benchmarks. One with the without the patch, one with it.
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:114]: "One with the without" -> "One without"

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+In the context of this project, we run the usual benchmarks sensitive to C++
+changes and we haven't identified any [regression][Perfherder] in term of
+performances.
+
+
+## Current status
+
+Firefox nightly on Linux is now compiled with the stack-clash-option from
+January 8st 2021. We have not detected any regressions since it landed.
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:130]: 8st -> 8th

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+If everything goes well, this change should ship with Firefox 86 (planned for
+mid february 2021).
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:132]: february -> February

File: content/posts/2021-01-05-stack-clash-protection.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+# Validation With Rust
+
+Rust has long support the callback style of the LLVM `probe-stack` attribute,
```
Comment (lattner) [inline content/posts/2021-01-05-stack-clash-protection.md:136]: support -> supported

**Review by lattner — COMMENTED** [review]
This is a really great post, thank you for writing this!  I noticed a few minor typos, but otherwise I think this is good to go!

**Pull request closed without merging**